### PR TITLE
Move to PeerJS

### PR DESCRIPTION
**Pull request by Nabel0721**
* Moves to using [PeerJS](https://peerjs.com/docs.html#start) for Peer to Peer communication instead of the previous system
* Reduces the size of the tokens that need to be copied, and made it so only the DM needs to send tokens to the players. (not the other way around)
* Should make it easier to later add support for player actions if needed.

*Does not currently work with the Local Connection button in the DM Screen. I'm looking into this, but right now I'm stumped.*

* Tests passed: 
  * ESLint
  * StyleLint
 
Failed test_all in linkcheck.  
(Example failure {
Missing link: {@creature Jelayne|AI} in file ./data/adventure/adventure-ai.json (evaluates to "bestiary.html#jelayne_ai")
Similar URLs were: []
}
I didn't touch these files, so I'm pretty sure I'm just missing files.)

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{
```
Comment (TheGiddyLimit) [inline .vscode/launch.json]: Please remove this file, and/or add `.vscode` to the `.gitignore`

File: js/dmscreen-initiativetracker.js [hunk 2]
```diff
@@ -43,19 +50,11 @@ class InitiativeTracker {
...
-			if (p2pMeta.serverInfo === null) return;
-
-			p2pMeta.rows = p2pMeta.rows.filter(r => !r.isDeleted);
-			p2pMeta.serverInfo = p2pMeta.serverInfo.filter(r => {
-				if (r.isDeleted) {
-					r.server.close();
-					return false;
-				} else return true;
-			});
+			if (srvPeer._connectionsArray.length < 1) return;
```
Comment (TheGiddyLimit) [inline js/dmscreen-initiativetracker.js]: Can be simplified to `if (!srvPeer._connectionsArray.length) return;`

File: js/dmscreen-playerinitiativetracker.js [hunk 3]
```diff
@@ -15,15 +15,15 @@ class InitiativeTrackerPlayer {
 		const view = new InitiativeTrackerPlayerMessageHandlerScreen();
 		view.setElements($meta, $head, $rows);
 
+		var ui = new InitiativeTrackerPlayerUi(view);
```
Comment (TheGiddyLimit) [inline js/dmscreen-playerinitiativetracker.js]: Please use `let`/`const` over `var` (always `const` where possible)

File: js/history.js [hunk 4]
```diff
@@ -31,7 +31,7 @@ class History {
 					const id = $el.attr("id");
 					History.lastLoadedId = id;
 					loadHash(id);
-					document.title = decodeURIComponent($el.attr("title")) + " - 5etools";
+					document.title = decodeURIComponent($el.attr("title")).concat(" - 5etools");
```
Comment (TheGiddyLimit) [inline js/history.js]: Template strings are preferred over `concat`, i.e.
```
`${decodeURIComponent($el.attr("title"))} - 5etools`
```

File: js/inittrackerplayerview.js [hunk 5]
```diff
@@ -6,17 +6,20 @@ window.addEventListener("load", () => {
 	ExcludeUtil.pInitialise(); // don't await, as this is only used for search
 
 	const view = new InitiativeTrackerPlayerMessageHandlerPage();
+	var ui = new InitiativeTrackerPlayerUi(view);
```
Comment (TheGiddyLimit) [inline js/inittrackerplayerview.js]: `const`/`let` over `var`

File: node/bestiary.js [hunk 6]
```diff
@@ -1,4 +1,4 @@
-const fs = require('fs');
+const fs = require("fs");
```
Comment (TheGiddyLimit) [inline node/bestiary.js]: You can just delete this file; it's not used. I need to force-sync this mirror with the main repo someday, essentially :grin:

File: node/spellcasting.js [hunk 7]
```diff
@@ -1,4 +1,4 @@
-const fs = require('fs');
```
Comment (TheGiddyLimit) [inline node/spellcasting.js]: This file can be deleted too

File: package.json [hunk 8]
```diff
@@ -47,7 +47,9 @@
 		"stylelint-config-sass-guidelines": "^6.1.0",
 		"xmlbuilder": "^13.0.2"
 	},
-	"dependencies": {},
+	"dependencies": {
```
Comment (TheGiddyLimit) [inline package.json]: We don't use node dependencies; please remove this. You've got the file committed in `lib/` anyway

**Comment by Nabel0721** [comment]
Should be good to go when you get the chance. Happy holidays!

**Comment by Nabel0721** [comment]
The conflicts in the dmscreen file look easy enough to fix, but I'm not sure if you want to keep the changes in utils-p2p.js or not. If so, I'm not sure where they'd go.

**Comment by TheGiddyLimit** [comment]
Ah, don't worry about it. I'll deal with it when I "merge" this to the main repo 👍

**Comment by Nabel0721** [comment]
I've been trying to figure out how to get the Local Connection button to work, and I just can't figure out where I'm screwing up. The connection is created on one end but not completed on the other, while the exact same operations work fine for a remote connection. I'm thinking we could just only do remote, since this would simplify the UI, and it's not a pain to copy the tokens back and forth anymore (you just grab the token you want from the right tracker and paste it in, rather than having a back and forth with large tokens). Thoughts?

**Comment by Nabel0721** [comment]
Or maybe instead of having two separate buttons for local/remote, just add a button to the UI to search for local trackers and if one is found use that to fill in the token value for the field. That way we aren't splitting up the call functionality, just the way we get the token.

**Comment by Nabel0721** [comment]
Nevermind, I figured out the local thing. There'll be another commit coming soon with that functionality added, as well as disabling connection buttons until the connector object is ready to go.

**Comment by TheGiddyLimit** [comment]
So, mysteriously, none of this works for me. Neither does the old implementation, though, so I assume there's something going on with my current setup (junky laptop and even junkier internet connection at my folks' place) which is killing it. Just to confirm, though, could you try opening:

`server.html`
```html
<html><body>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/peerjs/1.0.4/peerjs.min.js"></script>
<script>
	const peer = new Peer();
	peer.on('open', id => console.log('My peer ID is:', id));
	peer.on('connection', conn => conn.on('data', data => console.log(data)));
</script>
</body></html>
```

`client.html`
```html
<html><body>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/peerjs/1.0.4/peerjs.min.js"></script>
<script>
	const peer = new Peer();
	peer.on('open', id => console.log('My peer ID is:', id));
	function testConnect (id) {
		const conn = peer.connect(id);
		conn.on('open', () => conn.send('hi!'));
	}
</script>
</body></html>
```

and run `testConnect("id-from-server-console")` on the "client" page? For me, nothing happens; the "server" never receives the message. The code is that from peerjs' docs, albeit with some syntax shuffled around.

**Comment by Nabel0721** [comment]
Works for me. The server console outputs "hi!" after I run `testConnect("id-here") with the id from the server page.

**Comment by Nabel0721** [comment]
What browser are you using? I've tested mainly in firefox and chrome

**Comment by TheGiddyLimit** [comment]
🤔 good shout, it works on Firefox. I was using Chrome. Interesting. ~~Shame, since Firefox's dev tools are notably shittier.~~

**Comment by Nabel0721** [comment]
I'll test it in Chrome on my end. All my debugging was done with VSCode's chrome extensions, so it's weird to me that it wouldn't work on there for you.

**Comment by Nabel0721** [comment]
Worked fine in Chrome for me. Running version 79 in Ubuntu 19.10

**Comment by Nabel0721** [comment]
Tested the server/client files in Chromium as well. No issues.

**Comment by Nabel0721** [comment]
If you have questions, after like 2:00 EST today I can hop in Discord and work through this with you. Just let me know.

**Comment by TheGiddyLimit** [comment]
Apologies, stepped out for a bit.

Thanks for the thorough testing; downloaded Chromium and it seems fine also. No idea what's happened to my Chrome, but whatever. Not my main setup so I'll worry about that never. 

Cheers, I'll give you a shout if I need anything else. The code was making sense, so I hope that'll not be necessary :grin:

**Comment by Nabel0721** [comment]
If you see any code that doesn't need to be there anymore, feel free to nix it. I was too scared to touch some stuff. I'll probably keep working on cleaning this up over the next couple weeks, but I won't pester you with any more commits until this is done getting merged. 

I'm glad the code makes sense, I did my best to have flow of control be apparent.

**Comment by TheGiddyLimit** [comment]
And I _think_ that should be that. I would recommend closing this PR and branching fresh off `master` for any further tweaks; always feel bad clobbering people's GitHub contrib stats, but this repo's just a shitty mirror. You've been credited in the real commits in the main repo, fear not 😉

**Comment by Nabel0721** [comment]
Cool. When do you think I’ll see the changes reflected on the main site?

**Comment by TheGiddyLimit** [comment]
Soon:tm: Hopefully/probably within 24 hours from now.

**Comment by Nabel0721** [comment]
Cool. I’ve got Dnd sessions the week after next, and I’m excited to have this tool available. Having the DM control initiative and players have a view on their phones is really nice

**Comment by TheGiddyLimit** [comment]
It's live 🎉

**Pull request closed without merging**